### ISSUE TITLE
gcp-gce failed to request new Vault token

### ISSUE THREAD

**Issue by connorlwilkes** [bank-vaults/bank-vaults#1268]
**Describe the bug**:
Using the webhook vault secret injector when I try and authenticate with a gcp-gce service account I am getting back the following error:

`time="2021-02-22T18:24:26Z" level=error msg="failed to request new Vault token" app=vault-secrets-webhook err="Error making API request.\n\nURL: PUT https://<vault-url>/v1/auth/gcp/login\nCode: 500. Errors:\n\n* error converting input map[access_token:<access_token> expiry:2021-02-22T18:41:23.661415495Z token_type:Bearer] for field \"jwt\": '' expected type 'string', got unconvertible type 'map[string]interface {}'"`

**Expected behaviour**:
Authentication to be successful and configMap to be filled with data from vault

**Steps to reproduce the bug**:
Deploy a Vault 1.5.4+prem cluster, set up GCP auth, create configMap with annotation auth type of vault.security.banzaicloud.io/vault-auth-method: "gcp-gce". See error appear.

**Additional context**:
External non bank vaults vault: Vault 1.5.4+prem

**Environment details**:
- Kubernetes version (e.g. v1.10.2):
- Cloud-provider/provisioner GKE
- bank-vaults version (e.g. 0.4.17): master branch (webhook only)
- Install method - helm but external vault with webhook only installed from bank vaults
- Logs from the misbehaving component (and any other relevant logs): (see above log)
- Resource definition (possibly in YAML format) that caused the issue, without sensitive data:
`apiVersion: v1
kind: ConfigMap
metadata:
  name: test-configmap
  namespace: vault-injector-system
  annotations:
    vault.security.banzaicloud.io/vault-addr: "https://<vault-url>"
    vault.security.banzaicloud.io/vault-role: "<role>"
    vault.security.banzaicloud.io/vault-path: "gcp"
    vault.security.banzaicloud.io/vault-auth-method: "gcp-gce"
    vault.security.banzaicloud.io/inline-mutation: "true"
    vault.security.banzaicloud.io/vault-skip-verify: "true"
data:
  TEST_SECRET: vault:<path>/data/injector-test#hello`

/kind bug

**Comment by connorlwilkes** [comment]
Looks to me like:
`jwt, err := tokenSource.Token()` is called which returns as per https://pkg.go.dev/golang.org/x/oauth2@v0.0.0-20210220000619-9bb904979d93#Token a Token type which has various fields. This is then set to the jwt value: https://github.com/banzaicloud/bank-vaults/blob/master/pkg/sdk/vault/client.go#L386 so now effectively we have jwt as a Token type which has string fields and a time.Time field for expiry.

Then there is a call later on to get the token:
```
secret, err := logical.Write(fmt.Sprintf("auth/%s/login", o.authPath), loginData)
					if err != nil {
						client.logger.Error("failed to request new Vault token", map[string]interface{}{"err": err})
						time.Sleep(1 * time.Second)
						continue
					}
```
LoginData here is what is returned from the above where it is a map that looks as follows:
```
{
  "jwt": Token,
  "role": "roleString"
}
```
In vault, this gets unmarshalled as a Json object (https://github.com/hashicorp/vault/blob/b2927012ba9131f68606debec13bfc221b221912/vendor/github.com/hashicorp/vault/api/logical.go#L134) for the request that then looks like the following:
```
{
  "jwt": map[string],
  "role": "roleString"
}
```
This is wrong because JWT is expecting a signed JWT token string and not a map[string]. 

I guess this can be fixed by ensuring that the token is signed and that JWT is provided as a string. Can someone verify that I am correct?

**Comment by bonifaido** [comment]
Hi @connorlwilkes,
how is your PR (https://github.com/banzaicloud/bank-vaults/pull/1272) related to this issue, is it a fix for it?

**Comment by connorlwilkes** [comment]
Hi @bonifaido 

No my PR relates to enabling GCP-IAM auth which is slightly different from GCE auth (just expects different set of information in JWT). Originally I thought my issue was caused by the above so I did make inroads to fix it through generating a signed JWT etc however I realised that with WI in GKE it enforces GCP-IAM auth.

I can submit the change I made to fix the above but unfortunately I am currently not able to test it properly bar spinning up a whole new GKE cluster with no WI. I was wondering actually if anyone else in the community had tested the GCP-GCE auth firstly.

The change is effectively the following:

```
case GCPGCEAuthMethod:
				loginDataFunc = func() (map[string]interface{}, error) {
					c := metadata.NewClient(nil)
					queryString := fmt.Sprintf("instance/service-accounts/default/identity?audience=vault/%s", o.role)
					jwt, err := c.Get(queryString)
					if err != nil {
						return nil, err
					}

					return map[string]interface{}{
						"jwt":  jwt,
						"role": o.role,
					}, nil
				}
```
This is in line with https://www.vaultproject.io/docs/auth/gcp#gce.
